docs/source/plugins/developing_plugins.rst

```diff
@@ -0,0 +1,1810 @@
+For example, here's a simple example of collecting a single string input from
+the user:
+
+.. code-block:: python
+    :linenos:
+
+    def resolve_input(self, ctx):
+        inputs = types.Object()
+        inputs.str("message", label="Message", required=True)
+        return types.Property(inputs, view=types.View(label="Static example"))
+
+    def execute(self, ctx):
+        the_message = ctx.params["message"]
+
+If the :ref:`operator's config <operator-config>` declares `dynamic=True`, then
+:meth:`resolve_input() <fiftyone.operators.operator.Operator.resolve_input>`
+will be called after each user input, which allows you to construct dynamic
+forms whose components may contextually change based on the already provided
+values and any other aspects of the
+:ref:`execution context <operator-execution-context>`:
+
+.. code-block:: python
+    :linenos:
+
+    import fiftyone.brain as fob
+
+    def resolve_input(self, ctx):
+        inputs = types.Object()
+        brain_keys = ctx.dataset.list_brain_runs()
+
+        if not brain_keys:
+            warning = types.Warning(label="This dataset has no brain runs")
+            prop = inputs.view("warning", warning)
+            prop.invalid = True  # so form's `Execute` button is disabled
+            return
+
+        choices = types.DropdownView()
+        for brain_key in brain_keys:
+            choices.add_choice(brain_key, label=brain_key)
+
+        inputs.str(
+            "brain_key",
+            required=True,
+            label="Brain key",
+            description="Choose a brain key to use",
+            view=choices,
+        )
+
+        brain_key = ctx.params.get("brain_key", None)
+        if brain_key is None:
+            return  # single `brain_key`
+
+        info = ctx.dataset.get_brain_info(brain_key)
+
+        if isinstance(info.config, fob.SimilarityConfig):
+            # We found a similarity config; render some inputs specific to that
+            inputs.bool(
+                "upgrade",
+                label"Compute visualization",
+                description="Generate an embeddings visualization for this index?",
+                view=types.CheckboxView(),
+            )
+
+        return types.Property(inputs, view=types.View(label="Dynamic example"))
+
+.. note::
+
+    Avoid expensive computations in
+    :meth:`resolve_input() <fiftyone.operators.operator.Operator.resolve_input>`
+    or else the form may take too long to render, especially for dynamic inputs
+    where the method is called after every user input.
+
+Remember that properties automatically handle validation for you. So if you
+configure a property as `required=True` but the user has not provided a value,
+the property will automatically be marked as `invalid=True`. The operator's
+`Execute` button will be enabled if and only if all input properties are valid
+(recurisvely searching nested objects).
+
+.. note::
+
+    As the example above shows, you can manually set a property to invalid by
```

Comment:
This seems to be moved around or the example is missing that it refers to.

Reply:
It's referring to the code block in this section. I tweaked so that this note is as close as possible to the code block now.